### PR TITLE
DB backups without cron

### PR DESCRIPTION
**Pull request by Mart124**
Hi,

As in #601, let's get rid of `cron` for database backups.
`cron` was the last process running as root in Bitwarden.
In addition, we don't need it for a unique daily simple task.

Some small additional features :
- admin can disable these daily backups setting env `BACKUP_DB=0` ;
- `backup-db.sh` can still be executed manually to perform a backup whenever admin wants.

Thank you 👍

### CODE REVIEW

**Comment by Mart124** [comment]
Here also, @kspearrin, feel free to squash-merge to make things proper :)

File: util/MsSql/entrypoint.sh [hunk 1]
```diff
@@ -42,18 +42,17 @@ fi
...
 chown $USERNAME:$GROUPNAME /backup-db.sh
 chown $USERNAME:$GROUPNAME /backup-db.sql
 
-# Sounds like gosu keeps env when switching, but of course cron does not
-env > /etc/environment
-cron
+# Launch a loop to backup database on a daily basis
+if [ "$BACKUP_DB" != "0" ]
+then
+    su - bitwarden -s /bin/sh -c "/backup-db.sh loop >/dev/null 2>&1 &"
```
Comment (kspearrin) [inline util/MsSql/entrypoint.sh]: Why not gosu?

Comment (Mart124) [reply util/MsSql/entrypoint.sh]: U're right as it's the preferred method (over sudo / su), according to the docker's doc.
Thx 👍

**Comment by kspearrin** [comment]
@Mart124 I am having issues getting mssql container to start up now. I think it may hav something to do with this change since you are writing to the errorlog:

```
2020-01-22 20:38:32.92 Server      Error: 17049, Severity: 16, State: 1.
2020-01-22 20:38:32.92 Server      Unable to cycle error log file from '/var/opt/mssql/log/errorlog' to '/var/opt/mssql/log/errorlog.1' due to OS error '31(A device attached to the system is not functioning.)'. A process outside of SQL Server may be preventing SQL Server from reading the files. As a result, errorlog entries may be lost and it may not be possible to view some SQL Server errorlogs. Make sure no other processes have locked the file with write-only access."
2020-01-22 20:38:32.92 Server      Microsoft SQL Server 2017 (RTM-CU14) (KB4484710) - 14.0.3076.1 (X64)
        Mar 12 2019 19:29:19
        Copyright (C) 2017 Microsoft Corporation
        Express Edition (64-bit) on Linux (Ubuntu 16.04.6 LTS)
2020-01-22 20:38:32.93 Server      UTC adjustment: 0:00
2020-01-22 20:38:32.93 Server      (c) Microsoft Corporation.
2020-01-22 20:38:32.93 Server      All rights reserved.
2020-01-22 20:38:32.93 Server      Server process ID is 28.
2020-01-22 20:38:32.93 Server      Logging SQL Server messages in file '/var/opt/mssql/log/errorlog'.
2020-01-22 20:38:32.93 Server      Registry startup parameters:
         -d /var/opt/mssql/data/master.mdf
         -l /var/opt/mssql/data/mastlog.ldf
         -e /var/opt/mssql/log/errorlog
2020-01-22 20:38:32.94 Server      Error: 17114, Severity: 16, State: 1.
2020-01-22 20:38:32.94 Server      Error (null) occurred while opening file '/var/opt/mssql/data/master.mdf' to obtain configuration information at startup time. An invalid startup option might have caused the error. Verify your startup options, and correct or remove them if necessary.
```

**Comment by Mart124** [comment]
Let me dig into this @kspearrin.

**Comment by Mart124** [comment]
Strangely enough, we do dot write to the `/var/opt/mssql/log/errorlog` file.
We simply read it thanks to `grep` command, looking for `BACKUP DATABASE successfully` string.
If found we delete old backups.
Let me try to reproduce your issue.

**Comment by Mart124** [comment]
When issue occurs, what the following command, executed inside the `mssql` container, returns ?
`lsof /var/opt/mssql/log/errorlog`

**Comment by Mart124** [comment]
`A device attached to the system is not functioning.`
Looking at this log extract, sounds like it my be related to your underlying FS.
We see at the end of your log extract that it also fails to open `/var/opt/mssql/data/master.mdf`.
Sounds like a more global issue.

File: util/MsSql/backup-db.sh [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
-export now=${1:-$(date +%Y%m%d_%H%M%S)}
+while true
+do
+  # Sleep until next day
+  [ "$1" == "loop" ] && sleep $((24 * 3600 - (`date +%H` * 3600 + `date +%M` * 60 + `date +%S`)))
 
-# Do a new backup
-/opt/mssql-tools/bin/sqlcmd -S localhost -U sa -P ${SA_PASSWORD} -i /backup-db.sql
+  # Backup timestamp
+  export now=$(date +%Y%m%d_%H%M%S)
```
Comment (haneef95) [inline util/MsSql/backup-db.sh:9]: `${1:-` was a good add:

> One could do things like this (I do actually) :
`docker exec -ti d323a1c57728 /backup-db.sh $(date +%Y%m%d_%H%M%S)`
(in a more global backup plan for example)

https://github.com/bitwarden/server/pull/298/files#r191313662

Any reason why it was removed? Seems accidental to me.

Thanks

Comment (haneef95) [reply util/MsSql/backup-db.sh:9]: @Mart124

Comment (Mart124) [reply util/MsSql/backup-db.sh:9]: HI @haneef95, it has been removed now that you can set the container timezone properly, your backups will then be properly timestamped :
```
# cat docker-compose.override.yml 
version: '3'

services:
  mssql:
    environment:
      - TZ=Europe/Berlin
      - BACKUP_DB=0
```
Last line is an example to show you that you can also disable the built-in backup schedule itself.

Does it suit your needs ?

Comment (haneef95) [reply util/MsSql/backup-db.sh:9]: Thanks @Mart124, I see. And I also later realised that when executed from `entrypoint.sh` it'll likely name the file as `loop`:
https://github.com/bitwarden/server/blob/9f96e4ce90a24f9830bf1ab117ce3d3167d460a7/util/MsSql/entrypoint.sh#L62

So, I guess it had to be removed.

I've submitted a PR #1720 now however, with some additional features, including the ability to rename the backup file and/or give it a suffix. These should suit some of my needs and of course the community. Feel free to drop a line on that with your thoughts. 😄

**Pull request merged**